Terraform 0.12 upgrades release notes

evrardjp opened the pull request:
When upgrading to terraform 0.12 our users will need to upgrade their
terraform files and state in place. This will do the trick.

Closes: https://github.com/SUSE/avant-garde/issues/1116

Commit:
Terraform 0.12 upgrades release notes

When upgrading to terraform 0.12 our users will need to upgrade their
terraform files and state in place. This will do the trick.

Closes: https://github.com/SUSE/avant-garde/issues/1116

pablochacin commented on adoc/MAIN.release-notes.adoc:
```diff
@@ -19,6 +19,65 @@ Release notes usually only list changes that happened between two subsequent rel
+For this, enter your {tf} files/state folder and:
+
+* Migrate {tf} files with the automatic migration tool:
+  ``terraform 0.12upgrade``.
```
A link to the official documentation would be handy.

evrardjp replied:
Agreed.

evrardjp replied:
Done.

pablochacin commented on adoc/MAIN.release-notes.adoc:
```diff
@@ -19,6 +19,65 @@ Release notes usually only list changes that happened between two subsequent rel
+
+* Replace any boolean written as a number to false/true. For example, replace
+  ``wait_for_lease = 1 `` with ``wait_for_lease = true`` in
+  ``libvirt/{lb|master|worker}-instance.tf``
```
I'm not sure libvirt is currently released to customers.

pablochacin commented on adoc/MAIN.release-notes.adoc:
```diff
@@ -19,6 +19,65 @@ Release notes usually only list changes that happened between two subsequent rel
+* Migrate {tf} files with the automatic migration tool:
+  ``terraform 0.12upgrade``.
+
+==== Extra operations for in-place upgrade of libvirt terraform files
```
I'm not sure libvirt is currently released to customers.

evrardjp replied:
Ok, do you want to remove it?

evrardjp replied:
Removed.

pablochacin commented on adoc/MAIN.release-notes.adoc:
```diff
@@ -19,6 +19,65 @@ Release notes usually only list changes that happened between two subsequent rel
+
+* Migrate {tf} files with the automatic migration tool:
+  ``terraform 0.12upgrade``.
+
```
I think it would be helpful to enumerate briefly the kind of changes required, for example:
* boolean must be true/false
* all variables must be declared
* dependencies must be explicitly declared

pablochacin replied:
Also wondering why we don't provide a patch for the terraform files? Are the customers allowed to modify them?

evrardjp replied:
We can't assume they won't modify them :( 
They could literally copy the files from the up to date location too, like they would with a new cluster, plus add the lifecycle. But I am afraid they will mess something up, for example forgetting to add the lifecycle, or copying the versions.tf.

evrardjp replied:
> I think it would be helpful to enumerate briefly the kind of changes required, for example:
> 
>     * boolean must be true/false
> 
>     * all variables must be declared
> 
>     * dependencies must be explicitly declared

That's what I initially did, but it looked terrible, due to my poor adoc skills. Maybe I should have a preamble with these kind of things, and pointing to the tf 0.12upgrade cli.

evrardjp replied:
Added the explanations as a preamble.

pablochacin commented on adoc/MAIN.release-notes.adoc:
```diff
@@ -19,6 +19,65 @@ Release notes usually only list changes that happened between two subsequent rel
+  ``master-instance.tf`` and ``worker-instance.tf``. Add in those resources:
+  ``
+  lifecycle {
+    ignore_changes = [user_data]
```
I think this change is dangerous because will prevent updates of nodes if we ever change the userdata.

evrardjp replied:
Correct. It's dangerous. Should we make it clearer?

evrardjp replied:
Could you check if the latest version is good enough for you?

Itxaka replied:
afaik this is only for the update of the files no? should this be removed inmediatly after upgrading 0.11 to 0.12? or doing that means that your cluster will be recreated on the next terraform action?

Not clear to me if this is forever or temporary

evrardjp replied:
I think that if you remove it after the upgrade, the state machine would still have different user_data, which will lead to a complete teardown of the cluster. So yes it's a little bit forever.

@bergmannf can you confirm? I didn't test the multiple update scenario.

atighineanu commented:
Scenario 02 for TF 0.12.19 testing (openstack):
* Cloning skuba 
* Deploying on openstack lb1:m1:w1 cluster using terraform 0.11.14    PASSED
* Bootstrapping with skuba (from cloned repo) using specified infra folder in skuba PASSED
* Running skuba cluster status     PASSED

* Installing Terraform 0.12.19
* Fetching the #872 PR into skuba repository (git pull skuba; git fetch skuba pull/872/head:pr872; git checkout pr872)
---------------
* make new skuba
* Running skuba cluster status           PASSED
* Running terraform init (inside the /home/whoami/go/src/skuba/ci/infra/openstack/)    PASSED
* Running terraform refresh                 PASSED
* Running terraform output                  PASSED
* Running terraform destroy                PASSED
"""
Scenario 03 for TF 0.12.19 testing (openstack):
* Cloning skuba 
* Fetching the #872 PR into skuba repository (git pull skuba; git fetch skuba pull/872/head:pr872; git checkout pr872)
* Installing Terraform 0.12
* Bootstrapping with latest skuba  using specified infra folder in skuba          PASSED
* Running skuba cluster status           PASSED
* Running terraform init (inside the /home/whoami/go/src/skuba/ci/infra/openstack/)    PASSED
* Running terraform refresh                 PASSED
* Running terraform output                  PASSED
* Running terraform destroy                PASSED
===
Openstack looks fine...                        PASSED

evrardjp commented:
I will rebase today.

nkoranova committed:
Style and formatting edits.

Commit:
Add clarifications on files to change

Without this, it might not be clear on which files to apply the
patches.

nkoranova commented on adoc/MAIN.release-notes.adoc:
```diff
@@ -19,6 +19,82 @@ Release notes usually only list changes that happened between two subsequent rel
+* Migrate {tf} files with the automatic migration tool by running ``terraform 0.12upgrade``.
+* Run the extra operations for in-place upgrade, which follow just below.
+
+==== Extra Operations for In-place Upgrade of OpenStack {tf} Files
```
What about extra steps for VMWare? Has anyone tested this as well?

evrardjp replied:
No extra steps required.

nkoranova replied:
So then the VMWare steps are basically only up to this point:

> * Migrate {tf} files with the automatic migration tool by running ``terraform 0.12upgrade``.
?

Maybe it would make sense to also mention that, I'm not sure it's implicitly clear.

evrardjp replied:
I can add a "For OpenStack". Doing it now.

evrardjp replied:
Done.

nkoranova commented on adoc/MAIN.release-notes.adoc:
```diff
@@ -19,6 +19,82 @@ Release notes usually only list changes that happened between two subsequent rel
+==== Extra Operations for In-place Upgrade of OpenStack {tf} Files
+
+* Replace any boolean values written as a number with ``false``/``true``. For example, replace
+  ``default = 0`` with ``default = false`` in ``openstack/variables.tf``.
```
It would be more transparent, if we said which specific values are boolean and should be changed. We tell our users to deploy on openstack using `openstack/terraform.tfvars.example`, so which values would they change specifically if the deploy using that template?

evrardjp replied:
We can't presume things won't be modified, and that's why we are giving the instructions instead. 
dnsentry, cpi_enable, and *_vol_enabled are the ones in variables.tf by default. I can add extra text listing those, and mentioning it's a non exhaustive list?

nkoranova replied:
Yes, I would remove the `default` example and say they should change those three specifically plus any other ones they might have added.

evrardjp replied:
OK. Done.

Commit:
Add examples of boolean variables edition

This should clarify what needs to be done by the user.

Commit:
Clarify the extra steps are for OpenStack

Without this patch, one might wonder what to do with vmware, as
extra step. By default there is nothing to do. This should
clarify it.

Commit:
Remove cpi_enable from the list of vars to migrate

cpi_enable was already formatted as true/false. There is no need
to do anything there then.

Commit:
Remove reference to passwords

password was a variable introduced in beta, for convenience, and
was never documented or recommended.
Configuring password might trigger a user to define a password,
even though it's not necessary. This is naturally not recommended.

Without this patch, the user might want to define new fields,
which shouldn't be defined. This is a problem for security purposes.

This fixes it by removing the recommendation to define password.
If a pre-beta user is trying to use password, he will get a
warning, and re-evaluate. This is for the best.

Itxaka reviewed (approved):
latest changes :+1:

thehejik commented:
Hi, we miss a guide for migrating tf on vmware - this document is covering only openstack. Could you please enhance this doc for vmware as well?

evrardjp commented:
> Hi, we miss a guide for migrating tf on vmware - this document is covering only openstack. Could you please enhance this doc for vmware as well?

There are no extra steps for vmware that those mentioned, afaik.

thehejik commented:
> There are no extra steps for vmware that those mentioned, afaik.

Yes, it really looks like there are no extra steps needed - anyway could you please mention that in this doc? Something like - No extra steps needed on VMware apart `terraform 0.12upgrade --yes && terraform init`.

evrardjp commented:
Yes it was tough wording, so we decided to go for "For Openstack, ...." . I suppose adding an extra line "vmware doesn't need extra steps" would clarify.

Commit:
Clarifications about vmware

VMWare doesn't need extra operations.
However, people asked me if there were any operations required,
so I suppose that making it explicit would be better for
readability.

evrardjp commented:
Done.

Itxaka reviewed (approved):
latest change was not needed IMHO, but more info never hurt anybody....probably. Now Im not sure about that.

nkoranova commented:
I'd say the VMWare clarification is good,  it's better to say it explicitly and have "extra information" than none at all.

nkoranova commented:
Thanks for your contributions everyone! :bowing_woman:
